codec/src/codec.rs

```diff
@@ -3,53 +3,103 @@
 use crate::error::Error;
 use bytes::{Buf, BufMut, BytesMut};
 
-/// Trait for types that can be encoded to and decoded from bytes
-pub trait Codec: Sized {
-    /// Encodes this value to a writer.
+/// Trait for types that can be used as configuration during decoding.
+pub trait Config: Clone + Send + 'static {}
+
+// Automatically implement `Config` for matching types.
+impl<T: Clone + Send + 'static> Config for T {}
+
+/// Trait for types that can be written (encoded) to a buffer.
+pub trait Write {
+    /// Encodes this value by writing to a buffer.
+    ///
+    /// Implementations should panic if the buffer doesn't have enough capacity.
     fn write(&self, buf: &mut impl BufMut);
```

Comment:
Wondering if we should bit the bullet now and also integrate `cfg` with `write`? I suspect we'll "eventually" want this but not sure when.

Reply:
Let's omit it for now, I'm not convinced we should add this as it starts to add more complexity and additional type proliferation. I think it's standard to not use a config when encoding.